benchmark/main.c

```diff
@@ -0,0 +1,354 @@
+			char *end;
+			long c = strtol(opt.value, &end, 10);
+
+			if (c <= 0 || *end) {
+				fprintf(stderr, "%s: invalid count '%s'\n'", progname, opt.value);
+				print_usage(stderr);
+				error = 1;
+				goto done;
+			}
+
+			count = (unsigned int)c;
+		} else if (strcmp(opt.spec->name, "benchmark") == 0) {
+			benchmark_name = argv[parser.idx];
+		}
+	}
+
+	if ((error = benchmark_init(&benchmark, benchmark_name, cmd_args.length, (const char **)cmd_args.contents)) < 0) {
+		if (error == GITBENCH_EARGUMENTS)
+			error = 1;
+		goto done;
+	}
+
+	gitbench_try(benchmark_run(&runs, benchmark, count));
+
+	report_benchmark(&runs, benchmark);
+
+	if (profile_alloc)
+		report_alloc();
+
+done:
+	if (error < 0) {
+		const git_error *err = giterr_last();
+		fprintf(stderr, "%s: %s\n", progname, err ? err->message : "unknown error");
+	}
+
+	benchmark_free(&runs, benchmark);
+
+	gitbench_shutdown();
+	return error ? 1 : 0;
+}
+
+int gitbench_init(void)
+{
+	gitbench_alloc_init();
+	return git_libgit2_init();
+}
+
+void gitbench_shutdown(void)
+{
+	git_libgit2_shutdown();
+	gitbench_alloc_shutdown();
+}
+
+const char *basename(const char *in)
+{
+	const char *c, *last = in;
+	for (c = in; *c; c++)
+		if ((*c == '/' || *c == '\\') && *(c+1)) last = c+1;
+	return last;
+}
+
+int benchmark_init(gitbench_benchmark **out, const char *name, int argc, const char **argv)
+{
+	const gitbench_benchmark_spec *benchmark_spec = NULL;
+
+	if (!name) {
+		print_usage(stdout);
+		return GITBENCH_EARGUMENTS;
+	}
+
+	if ((benchmark_spec = benchmark_spec_lookup(name)) == NULL) {
+		fprintf(stderr, "%s: unknown benchmark '%s'\n", progname, name);
+		print_usage(stderr);
+		return GITBENCH_EARGUMENTS;
+	}
+
+	return benchmark_spec->init(out, argc, argv);
+}
+
+const gitbench_benchmark_spec *benchmark_spec_lookup(const char *name)
+{
+	const gitbench_benchmark_spec *spec;
+
+	for (spec = gitbench_benchmarks; spec->name; spec++) {
+		if (strcmp(spec->name, name) == 0)
+			return spec;
+	}
+
+	return NULL;
+}
+
+int benchmark_run(
+	git_vector *runs,
+	gitbench_benchmark *benchmark,
+	size_t count)
+{
+	size_t i;
+	int error = 0;
+
+	assert(runs && benchmark);
+
+	if (git_vector_init(runs, count, NULL) < 0)
+		return -1;
+
+	for (i = 0; i < count; i++) {
+		gitbench_run *run;
+
+		if (gitbench_run_init(
+			&run, i+1,
+			benchmark->operation_cnt,
+			benchmark->operations) < 0)
+			return -1;
+
+		run->verbosity = verbosity;
+		run->profile_alloc = profile_alloc;
+
+		if (git_vector_insert(runs, run) < 0)
+			return -1;
+	}
+
+	for (i = 0; i < count; i++) {
+		gitbench_run *run = git_vector_get(runs, i);
+
+		if ((error = gitbench_run_start(run)) < 0 ||
+			(error = benchmark->run_fn(benchmark, run)) < 0 ||
+			(error = gitbench_run_finish(run)) < 0)
+			break;
+	}
+
+	return error;
+}
+
+static void report_run(const char *desc, size_t num,
+	double execute, double setup, double cleanup, double total)
+{
+	if (desc)
+		printf("%s", desc);
```

Comment:
I would change all of the printf() calls that write results to an optional logfile.  Nothing worse than running a test in VS and have the output disappear when the program exits.